src/guiguts/mainwindow.py

```diff
@@ -407,13 +503,232 @@ def get_text(self) -> str:
-    # pos = super().mark_previous(index)
-    # return pos if pos else ""
+        Args:
+            IndexRange containing corners of block to be selected."""
+        self.tag_remove("sel", "1.0", tk.END)
+        min_row = min(col_range.start.row, col_range.end.row)
+        max_row = max(col_range.start.row, col_range.end.row)
+        min_col = min(col_range.start.col, col_range.end.col)
+        max_col = max(col_range.start.col, col_range.end.col)
+        for line in range(min_row, max_row + 1):
+            beg = IndexRowCol(line, min_col).index()
+            end = IndexRowCol(line, max_col).index()
+            # If line is too short for any text to be selected, select to
+            # beginning of next line (just captures the newline). This
+            # is then dealt with in column_copy_cut.
+            if self.get(beg, end) == "":
+                end += "+ 1l linestart"
+            self.tag_add("sel", beg, end)
+
+    def selected_ranges(self) -> list[IndexRange]:
+        """Get the ranges of text marked with the `sel` tag.
+
+        Returns:
+            List of IndexRange objects indicating the selected range(s)
+        """
+        ranges = self.tag_ranges("sel")
+        assert len(ranges) % 2 == 0
+        sel_ranges = []
+        for idx in range(0, len(ranges), 2):
+            sel_ranges.append(IndexRange(str(ranges[idx]), str(ranges[idx + 1])))
+        return sel_ranges
+
+    def column_copy_cut(self, cut: bool = False) -> None:
+        """Copy or cut the selected text to the clipboard.
+
+        A newline character is inserted between each line.
+
+        Args:
+            cut: True if cut is required, defaults to False (copy)
+        """
+        if not (ranges := self.selected_ranges()):
+            return
+        self.clipboard_clear()
+        for range in ranges:
+            start = range.start.index()
+            end = range.end.index()
+            # Trap case where line was too short to select any text
+            # (in do_column_select) and so selection-end was adjusted to
+            # start of next line. Adjust it back again.
+            if range.end.row > range.start.row and range.end.col == 0:
+                end += "- 1l lineend"
+            string = self.get(start, end)
+            if cut:
+                self.delete(start, end)
+            self.clipboard_append(string + "\n")
+
+    def column_paste(self) -> None:
+        """Paste the clipboard text column-wise, overwriting any selected text.
+
+        If more lines in clipboard than selected ranges, remaining lines will be
+        inserted in the same column in lines below. If more selected ranges than
+        clipboard lines, clipboard will be repeated until all selected ranges are
+        replaced.
+        """
+        # Trap empty clipboard or no STRING representation
+        try:
+            clipboard = self.clipboard_get()
+        except tk.TclError:
```

Comment:
Do we know what conditions this happens in? Does it only happen when the clipboard is empty? Feels like we're swallowing an error and going to silently fail at the paste.

Reply:
It fails if the clipboard is empty, or if the clipboard does not have a "STRING" format, e.g. it's an image. There doesn't appear to be a way to query the clipboard that won't generate a `tk.TclError`. There's a similar check in the pasting function in GG1:
```
    eval { $clipboard_text = $w->SelectionGet( -selection => "CLIPBOARD" ); };
    return unless ( defined $clipboard_text and length $clipboard_text );
    ...
    go on to paste clipboard contents
```